src/MongoDB.Driver.Core/Core/Compression/SnappyCompressor.cs

```diff
@@ -27,28 +27,89 @@ internal class SnappyCompressor : ICompressor
+        /// <paramref name="output"/>.
+        /// </summary>
+        /// <param name="input"> The input stream. Stream MUST HAVE read the exactly up to and including the
+        /// originalOpCode.</param>
+        /// <param name="output">The output stream. Stream MUST HAVE written the header exactly up to and including
+        /// compressorType aka compressorId.</param>
+        /// <remarks>
+        /// The precondition for <paramref name="input"/> mirrors the precondition for <see cref="Decompress"/>'s output
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Decompress"/>'s input
+        /// precondition.
+        /// </remarks>
         public void Compress(Stream input, Stream output)
         {
 #if NET452 || NETSTANDARD2_0
-            using (var snappyStream = new SnappyStream(output, CompressionMode.Compress, true))
-            {
-                input.EfficientCopyTo(snappyStream);
-            }
+            // extract uncompressedSize from OP_COMPRESSED header in output
+            const int uncompressedSizeOffset = -(UncompressedSizeSize + CompressorTypeSize);
+            var position = output.Position;
+            output.Seek(uncompressedSizeOffset, SeekOrigin.Current); // rewind head
+            var uncompressedSize = ReadInt32(output);
+            output.Position = position; // return head to original position
+
+            var uncompressedMessage = new byte[uncompressedSize]; // does not include uncompressed message headers
+            input.ReadBytes(uncompressedMessage, offset: 0, count: uncompressedSize, CancellationToken.None);
+            var maxSize = SnappyCodec.GetMaxCompressedLength(uncompressedSize);
+            var compressedBytes = new byte[maxSize];
+            var compressedSize = SnappyCodec.Compress(uncompressedMessage, 0, uncompressedSize, compressedBytes, 0);
+            output.Write(compressedBytes, 0, compressedSize);
 #else
             throw new NotSupportedException();
 #endif
         }
 
+        /// <summary>
+        /// Decompresses the message in <paramref name="input"/>, writing the uncompressed message to
+        /// <paramref name="output"/>.
+        /// </summary>
+        /// <param name="input">The input stream. Stream MUST HAVE read the header exactly up to and including
+        /// the compressorType aka CompressorId</param>
+        /// <param name="output">The output stream. Stream MUST HAVE written the message exactly up to and including
+        /// originalOpCode.</param>
+        /// <remarks>
+        /// The precondition for <paramref name="input"/> mirrors the precondition for <see cref="Compress"/>'s output
+        /// precondition.
+        /// The precondition for <paramref name="output"/> mirrors the precondition for <see cref="Compress"/>'s input
+        /// precondition.
+        /// </remarks>
```

Comment:
These comments might no longer be correct depending on what changes you make below.

Reply:
Applied changes as discussed offline!